realm/realm-library/src/main/cpp/jni_util/log.cpp

```diff
@@ -0,0 +1,189 @@
+{
+}
+
+JavaLogger::JavaLogger(JNIEnv* env, jobject java_logger)
+    :JniLogger(true)
+{
+    jint ret = env->GetJavaVM(&m_jvm);
+    if (ret != 0) {
+        throw std::runtime_error(util::format("Failed to get Java vm. Error: %d", ret));
+    }
+    m_java_logger = env->NewGlobalRef(java_logger);
+    jclass cls = env->GetObjectClass(m_java_logger);
+    m_log_method = env->GetMethodID(cls, "log", "(ILjava/lang/String;Ljava/lang/Throwable;Ljava/lang/String;)V");
+}
+
+JavaLogger::~JavaLogger()
+{
+    get_current_env()->DeleteGlobalRef(m_java_logger);
+}
+
+void JavaLogger::log(Log::Level level, const char* tag, jthrowable throwable, const char*,
+        const char* message)
+{
+    JNIEnv *env = get_current_env();
+    env->CallVoidMethod(m_java_logger, m_log_method, level, env->NewStringUTF(tag),
+            throwable, env->NewStringUTF(message));
+}
+
+bool JavaLogger::is_same_object(JNIEnv* env, jobject java_logger)
+{
+    return env->IsSameObject(m_java_logger, java_logger);
+}
+
+Log::Log()
+    : m_loggers()
+{
+    add_logger(get_default_logger());
+}
+
+Log& Log::shared()
+{
+    static Log log;
+    return log;
+}
+
+void Log::add_java_logger(JNIEnv* env, const jobject java_logger)
+{
+    std::shared_ptr<JniLogger> logger = std::make_shared<JavaLogger>(env, java_logger);
+    add_logger(logger);
+}
+
+void Log::remove_java_logger(JNIEnv* env, const jobject java_logger)
+{
+    std::lock_guard<std::mutex> lock(m_mutex);
+    for (auto it = m_loggers.begin(); it != m_loggers.end(); ++it) {
+        if ((*it)->m_is_java_logger &&
+                std::static_pointer_cast<JavaLogger>(*it)->is_same_object(env, java_logger)) {
+            m_loggers.erase(it);
+            break;
+        }
+    }
+}
+
+void Log::add_logger(std::shared_ptr<JniLogger> logger)
+{
+    std::lock_guard<std::mutex> lock(m_mutex);
+    m_loggers.push_back(logger);
+}
+
+void Log::remove_logger(std::shared_ptr<JniLogger> logger)
+{
+    std::lock_guard<std::mutex> lock(m_mutex);
+    for (auto it = m_loggers.begin(); it != m_loggers.end(); ++it) {
+        if (*it == logger) {
+            m_loggers.erase(it);
+        }
+    }
+}
+
+void Log::clear_loggers()
+{
+    std::lock_guard<std::mutex> lock(m_mutex);
+    m_loggers.clear();
+}
+
+void Log::set_level(Level level)
+{
+    s_level = level;
+}
+
+void Log::log(Level level, const char* tag, jthrowable throwable, const char* stacktrace,
+        const char* message)
+{
+    if (s_level <= level) {
+        std::lock_guard<std::mutex> lock(m_mutex);
+        for (auto& logger : m_loggers) {
+            logger->log(level, tag, throwable, stacktrace, message);
+        }
+    }
+}
+
+void CoreLoggerBridge::do_log(realm::util::Logger::Level level, std::string msg)
+{
+    // Ignore the leve threshold from the root logger.
+    Log::Level jni_level;
+    switch (level) {
+        case Level::trace: jni_level = Log::trace; break;
+        case Level::debug: // Map to same debug level.
```

Comment:
don't we need a `break` or we will fall down to `default`?
same for `all` & `off`